coredocument/read_acls.go

```diff
@@ -0,0 +1,56 @@
+package coredocument
+
+import (
+	"github.com/centrifuge/centrifuge-protobufs/gen/go/coredocument"
+	"github.com/centrifuge/go-centrifuge/errors"
+	"github.com/centrifuge/go-centrifuge/identity"
+)
+
+// ErrZeroCollaborators error when no collaborators are passed
+const ErrZeroCollaborators = errors.Error("require atleast one collaborator.")
+
+// initReadRules initiates the read rules for a given coredocument.
+// Collaborators are given Read_Sign action.
+// if the rules are created already, this is a no-op.
+func initReadRules(cd *coredocumentpb.CoreDocument, collabs []identity.CentID) error {
+	if len(cd.Roles) > 0 && len(cd.ReadRules) > 0 {
+		return nil
+	}
+
+	if len(collabs) < 1 {
+		return ErrZeroCollaborators
+	}
+
+	addCollaboratorsToReadRules(cd, collabs)
+	return nil
+}
+
+func addCollaboratorsToReadRules(cd *coredocumentpb.CoreDocument, collabs []identity.CentID) {
+	if len(collabs) == 0 {
+		return
+	}
+
+	// create a role for given collaborators
+	role := new(coredocumentpb.Role)
+	for _, c := range collabs {
+		c := c
+		role.Collaborators = append(role.Collaborators, c[:])
+	}
+
+	// add the role to Roles
+	cd.Roles = appendRole(cd.Roles, role)
```

Comment:
If this is an update or new version of the document wouldn't this appendRole append duplicates to the existing Roles of the previous version? I guess that depends on how we "prepare" the new version.
I can see that the initReadRules returns if there are rules/roles already. Will the `addCollaboratorsToReadRules` be only used from the `initReadRules` call? Otherwise I think it will add duplicates.